### ISSUE TITLE
Column comparison operator in exists clause

### ISSUE THREAD

**Issue by DanRymo** [hasura/graphql-engine#9446]
### Version Information
hasura/graphql-engine:v2.19.0

Server Version:
CLI Version 2.19.0:

### Environment
Running in Heroku
services:
  postgres:
    image: postgres:13.4
  graphql-engine:
    image: hasura/graphql-engine:v2.19.0

### What is the current behaviour?
First of all: Sorry if this explanation is messy and hard to follow, I found it very hard to describe.
The issue arises when trying to use the _not and _exists filters to deny access to specific rows in the org_stuff table (in the given example), based on the presence of certain rows in the same table.
Generally, it looks like _not in front of _exist doesn't just flip the result of _exist, but something more, this It's hard to use _exist to deny access in permissions.
Explicitly, I'm trying to make it so that a member of an organization can see public rows in org_stuff ONLY IF the organization the user is a member of ISN'T related to row(s)  in org_stuff with that stuff_type_id.

### What is the expected behavior?
Generally, I expect _exist to function as a boolean, being true if a row with the given argument(s) is found, else false. Placing a _not in front of it, I expect to flip the result of the _exist.
Instead of only denying access to the specific rows, the filter ends up denying access to all rows in the org_stuff table, even if the _exist filter doesn't find anything. To me this hints that _not in front of _exist does more than flip the result OR potentially _exist doesn't work as expected when used for permissions in the same table _exist is checking.

### How to reproduce the issue?
In this example, the issue occurs when setting up permissions for org_stuff. As I described above I want to make it so that a user, who is a member of an organization, cannot see other organizations org_stuff if the organization the user is a member of has a row in org_stuff with the same value for stuff_type_id.
So my setup is with the following tables where the relevant columns are defined in relationship with each other:
**organization:**
- id
- name

**user:**
- id
- name
- data

**organization_member:**
- id
- organization_id (relation to organization.id)
- member_type
- user_id (relation to user.id)

**org_stuff_type**
- id
- name
- description

**org_stuff:**
- id
- organization_id (relation to organization.id)
- stuff_type_id (relation to org_stuff_type.id)
- data
- public (boolean)

JSON for select permission for user in the org_stuff table:
```
{
    "_and": [
        { "public": { "_eq": true } },
        {
        "_not": {
            "_exists": {
            "_table": { "name": "org_stuff", "schema": "public" },
            "_where": {
                "_and": [
                {
                    "organization": {
                    "organization_members": {
                        "user_id": { "_eq": "X-Hasura-User-Id" }
                    }
                    }
                },
                { "type": { "_ceq": ["$", "type"] } }
                ]
            }
            }
        }
        }
    ]
}
```

### Please provide any traces or logs that could help here.

Sorry. none. Would love to be able to step trough and see where this fails.

### Any possible solutions/workarounds you're aware of?

Nope

### Keywords
permissions, _not, _exists, filtering

**Comment by BenoitRanque** [comment]
The logic is not being expressed correctly.

You're expecting, roughly:
```sql
SELECT *
FROM org_stuff os1
WHERE public = true &&
NOT EXISTS (
  SELECT 1
  FROM org_stuff os2
  WHERE EXISTS (
    SELECT 1 FROM organization_members om
    WHERE om.organization_id = os2.organization_id AND om.user_id = 'X-HASURA-USER-ID'
  )
  AND os2.type = os1.type
)
```

Instead you're getting:

```sql
SELECT *
FROM org_stuff os1
WHERE public = true &&
NOT EXISTS (
  SELECT 1
  FROM org_stuff os2
  WHERE EXISTS (
    SELECT 1 FROM organization_members om
    WHERE om.organization_id = os2.organization_id AND om.user_id = 'X-HASURA-USER-ID'
  )
  AND os2.type = os2.type
)
```

Note the subtle difference. You're comparing the type column of the org_stuff table in the exist check against itself, instead of comparing against the value outside the exists check.

However you're going to the root for your column comparison, so it's possible this behavior is a bug.

**Comment by DanRymo** [comment]
Thanks for your reply @BenoitRanque!
I'll test your SQL when I get back in office.
I think you're right -  the column comparison to root might be the issue as it explains at least some of  the behavior I see. But, when I test this permission inverted by removing the _not, the _exists column root comparison seems to work.

**Comment by BenoitRanque** [comment]
I can confirm that removing the `not` operator does not result in the behavior you expect, but the predicate will always be true so the results may look correct.

Here's a truncated version of the generated SQL with the `not` clause removed:

```sql
SELECT *
FROM "public"."org_stuff"
WHERE EXISTS (
    SELECT 1
    FROM "public"."org_stuff"
    WHERE ("public"."org_stuff"."type") = ("public"."org_stuff"."type")
  )
```

And with the `not` clause:

```sql
SELECT *
FROM "public"."org_stuff"
WHERE NOT EXISTS (
    SELECT 1
    FROM "public"."org_stuff"
    WHERE ("public"."org_stuff"."type") = ("public"."org_stuff"."type")
  )
```

Note I've reformatted the code, removed a bunch of parentheses, and removed most other components of the where clause.

Mostly the issue here is that the root reference to `"public"."org_stuff"` used is not aliased, so we can't reference it inside the subquery.

**Comment by DanRymo** [comment]
Ok, I see.
Is this something that will be fixed - and if yes, when?
I suppose I could implement this myself in an SQL function or make a view, but it would be a little bit more time demanding, plus require changes in front-end. Any other suggestions for implementation while waiting?

**Comment by BenoitRanque** [comment]
@DanRymo it seems the issue only really arises because your exists clause just so happens to have the current table as root.
So you _should_ be able to work around this by using the organization table as the root of your exists clause instead, and navigating relationships from there...

But unfortunately this runs into a UI bug as described in #9456

**Comment by DanRymo** [comment]
> @DanRymo it seems the issue only really arises because your exists clause just so happens to have the current table as root. So you _should_ be able to work around this by using the organization table as the root of your exists clause instead, and navigating relationships from there...
> 
> But unfortunately this runs into a UI bug as described in #9456

Yes, I tried that before opening this issue and found the same, and thought maybe I found an illegal reference (race condition sort of, referencing something before definition in the query... not exactly but sort of, if you get what I mean).
So again, any timeline on when a solution might be implemented?

**Comment by BenoitRanque** [comment]
This is a UI bug, and we're expecting the resolution to be quite fast, you can follow #9456 for updates

For the possible bug in the generated SQL, assuming it's confirmed as a bug, it would take much longer to fix as it's a more delicate issue.

**Comment by ejkkan** [comment]
Thanks for reporting the issue. We have a fix in place that will be part of one of our upcoming releases.

**Comment by DanRymo** [comment]
> Thanks for reporting the issue. We have a fix in place that will be part of one of our upcoming releases.
@ejkkan 
Is this fix implemented already? I ask since I haven't seen it specifically mentioned in a release note and it's a must have update for us.

**Comment by ejkkan** [comment]
It seems we missed adding a changelog entry for this fix.
It should be out with the latest release(2.21.x). Can you try that version out to see if it resolves the issue?

**Comment by DanRymo** [comment]
Thank you.
Seems to work fine while testing locally.
I'll tag you in a new comment if I come across it again during more extensive testing.
